frontend/src/api/Socket.ts

```diff
@@ -1,54 +1,92 @@
+/* eslint-disable consistent-return */
 import SockJS from 'sockjs-client';
 import Stomp, { Client, Message } from 'stompjs';
 
 let stompClient: Client;
 
-type User = {
+export type User = {
   nickname: string;
 }
 
 // Variable to hold the current connected state.
-let connected = false;
+let connected: boolean = false;
 
 // Create constants for the connection, subscription, and send message URLs.
 export const SOCKET_ENDPOINT:string = '/api/v1/socket/join-room-endpoint';
 const SUBSCRIBE_URL:string = '/api/v1/socket/subscribe-user';
-const SEND_USER_URL:string = '/api/v1/socket/user';
+const ADD_USER_URL:string = '/api/v1/socket/add-user';
+const DELETE_USER_URL:string = '/api/v1/socket/delete-user';
 
-export const sendUser = (nickname:string) => {
-  if (connected) {
-    stompClient.send(SEND_USER_URL, {}, nickname || 'Anonymous');
+export const isValidNickname = (nickname: string) => nickname.length > 0
+  && !nickname.includes(' ') && nickname.length <= 16;
+
+/**
+ * Add the user by sending a message via socket.
+ * @returns error, if present
+*/
+export const addUser = (nickname:string): Error | undefined => {
+  if (connected && isValidNickname(nickname)) {
+    stompClient.send(ADD_USER_URL, {}, nickname);
+  } else if (!connected) {
+    return new Error('The socket is not connected.');
   } else {
-    console.error('You must be connected to a socket before sending user information.');
+    return new Error('The provided nickname is invalid.');
   }
 };
 
-export const connect = (endpoint:string, nickname:string) => {
-  if (!connected) {
-    // Connect to given endpoint, subscribe to future messages, and send user message.
-    const socket: WebSocket = new SockJS(endpoint);
-    stompClient = Stomp.over(socket);
-    stompClient.connect({}, () => {
-      stompClient.subscribe(SUBSCRIBE_URL, (user: Message) => {
-        const userObject:User = JSON.parse(user.body);
-        console.log(`Welcome ${userObject.nickname} to the page!`);
-      });
-      sendUser(nickname);
-    });
-    // Reassign connected variable.
-    connected = true;
+/**
+ * Delete the user by sending a message via socket.
+ * @returns error, if present
+*/
+export const deleteUser = (nickname:string): Error | undefined => {
+  if (connected && isValidNickname(nickname)) {
+    stompClient.send(DELETE_USER_URL, {}, nickname);
+  } else if (!connected) {
+    return new Error('The socket is not connected.');
   } else {
-    console.error('You are already connected to a socket.');
+    return new Error('The provided nickname is invalid.');
   }
 };
 
-export const disconnect = () => {
+/**
+ * Connect and subscribe the user via socket.
+ * @returns error, if present
+*/
+export const connect = (endpoint:string, nickname:string,
+  setUsers: React.Dispatch<React.SetStateAction<User[]>>):
```

Comment:
Not too sure about the recommended practice of passing a React set state function. Instead of this, is it possible to return the users and call setUsers in Join.tsx in a then block?

Reply:
I actually do this in the next PR! So I'm happy to do it in this as well, I agree that is better practice. I will note though that in my next PR I add more `setX` functions, and I'm not sure how to structure it otherwise -- see [here](https://github.com/rocketden/main/blob/ed5b3253a252e58548e0bcfb8c9ed50c55feedbc/frontend/src/views/Join.tsx#L42). We can talk about this during our standup.

Reply:
After our standup, I have replaced this with a callback that is triggered every time the user's subscription receives a message; thus, I don't need to pass the any `setX` function.